users/app.py

```diff
@@ -37,20 +41,60 @@ def get_authorization():
     return jsonify(edit_access=authorized)
 
 
-@app.route('/v1/', methods=['PUT'])
-def add_update_user():
-    """Add or update the user in the db and returns new user object."""
-    user = request.getJSON()
-    if user is None:
-        # TODO(mukobi) validate the user object has everything it needs
-        return jsonify(error="You must supply a valid user in the body")
-    # TODO(mukobi) add or update the user in the database
-    added_new_user = True
-    # TODO(mukobi) get the new user object from the db to return
-    user_object = {
-        "user_id": "0", "username": "Dummy User", "edit_access": False
-    }
-    return jsonify(user_object), (201 if added_new_user else 200)
+@app.route('/v1/authenticate', methods=['POST'])
+def authenticate_user():
+    """Authenticate user, upsert in the db, and return new user object.
+
+    Request data:
+        'gauth_token': Google ID token to authenticate.
+
+    Response:
+        201: user object as it is in the db if authentication
+            was successful.
+        400: error message if authentication was not successful.
+    """
+    gauth_token = request.form.get('gauth_token')
+    if gauth_token is None:
+        return "Error: You must authenticate through Google.", 400
+    try:
+        idinfo = get_user_from_gauth_token(gauth_token)
+        user_object = {
+            "user_id": idinfo["sub"],
+            "name": idinfo["name"],
+            "is_organizer": False}  # authorization defaults false
+        upsert_user_in_db(user_object, app.config["COLLECTION"])
+        response = make_response(jsonify(user_object), 201)
+        response.headers['Access-Control-Allow-Origin'] = '*'
+        return response
+    except (AttributeError, ValueError, KeyError) as error:
+        return f"Error: {error}", 400
+
+
+def get_user_from_gauth_token(gauth_token):
+    """Validate the Google auth token and return it's user object.
+
+    Args:
+        gauth_token (str): Google ID token to authenticate.
+
+    Returns:
+        dict: User object. Contains attributes including:
+            'sub': Unique user ID.
+            'name': Full name of user.
+            'email': Email of user.
+            ...
+
+    Raises:
+        ValueError: Token is invalid.
+    """
+    # Authenticate token and match to client ID
+    idinfo = id_token.verify_oauth2_token(
+        gauth_token, requests.Request(), app.config["GAUTH_CLIENT_ID"])
+
+    if idinfo['iss'] not in ['accounts.google.com', 'https://accounts.google.com']:
+        raise ValueError('Wrong authentication token issuer.')
```

Comment:
Invalid instead of Wrong. Maybe return what the invalid value was? Is that safe to do? I don't know enough about authentication stuff.

Reply:
Done https://github.com/knative-portability/large-events/pull/31/commits/f2a8ccb636eee51d1e0c0a367c29b136ddf9486e. I think the token issuer is safe to return since it reflects the domain where the token was created and is simply checked against accounts.google.com